src/external/wannierio.jl

```diff
@@ -0,0 +1,693 @@
+    basis::PlaneWaveBasis, centers::AbstractVector{<:AbstractVector},
+    l::AbstractVector{<:Integer};
+    n::AbstractVector{<:Integer}=ones(Integer, length(centers)),
+    α::AbstractVector{<:Real}=ones(Real, length(centers))
+)
+    @assert length(centers) == length(l) == length(n) == length(α)
+    @assert all(length(c) == 3 for c in centers)
+    @assert all(n .> 0)
+    @assert all(α .> 0)
+
+    # I generate a fake `PspUpf` so as to reuse `build_projection_vectors_pswfcs`
+    labels = collect(zip(n, l, α))
+    psp_groups = [findall(Ref(lab) .== labels) for lab in Set(labels)]
+    psps = [radial_hydrogenic_upf(labels[first(group)]...) for group in psp_groups]
+    psp_positions = [centers[group] for group in psp_groups]
+
+    # note that on output the order of initial projections are different from
+    # the input order of (centers, l, n, α), due to reordering of psps
+    # output is a matrix of size (n_G, n_wann)
+    guess(kpt) = build_projection_vectors_pswfcs(basis, kpt, psps, psp_positions)
+
+    # we permute the order so that it matches the user input
+    # note the outputs also contain `m` angular momentum
+    perm = Int[]
+    for group in psp_groups
+        for g in group
+            start = sum(2 * l[i] + 1 for i in 1:(g-1); init=0)
+            append!(perm, start .+ (1:(2 * l[g] + 1)))
+        end
+    end
+    guess_perm(kpt) = guess(kpt)[:, perm]
+
+    # Lowdin orthonormalization
+    ortho_lowdin ∘ guess_perm
+end
+
+@timing function guess_amn_hydrogenic(
+    basis::PlaneWaveBasis, centers::AbstractVector{<:AbstractVector},
+    l::AbstractVector{<:Integer}, m::AbstractVector{<:Integer};
+    n::AbstractVector{<:Integer}=ones(Integer, length(centers)),
+    α::AbstractVector{<:Real}=ones(Real, length(centers))
+)
+    @assert length(centers) == length(l) == length(m) == length(n) == length(α)
+    @assert all(-l .<= m .<= l)
+
+    # find the indices of specified `l, m, n, α` in the output of
+    # `guess_amn_hydrogenic`, which computes all the `m`.
+    # e.g., with input `l=[0, 1, 1], m=[0, 1, -1]`, the returned orbitals
+    # of `guess_amn_hydrogenic` are
+    # `(l=0, m=0), (l=1, m=-1), (l=1, m=0), (l=1, m=1)`, so the computed
+    # `indices = [1, 4, 2]`.
+    indices = Int[]
+    lnα = []
+    for (ll, mm, nn, αα) in zip(l, m, n, α)
+        q = (ll, nn, αα)
+        !(q in lnα) && push!(lnα, q)
+        iq = findfirst(Ref(q) .== lnα)
+        lq = lnα[iq][1]
+        @info mm, lq
+        im = findfirst(mm .== (-lq:lq))
+        start = sum(2 * lnα[i][1] + 1 for i in 1:(iq-1); init=0)
+        push!(indices, start + im)
+    end
+
+    l = [q[1] for q in lnα]
+    n = [q[2] for q in lnα]
+    α = [q[3] for q in lnα]
+    guess = guess_amn_hydrogenic(basis, centers, l; n, α)
+    guess_filtered(kpt) = guess(kpt)[:, indices]
+    guess_filtered
+end
+
+"""
+Random Gaussian initial guess for Wannier functions.
+"""
+@timing function guess_amn_random(basis::PlaneWaveBasis, n_wann::Integer)
```

Comment:
in general things like `_random`, `_hydrogenic` should be done by dispatch on a marker struct in julia.